Dockerfile

```diff
@@ -245,32 +249,26 @@ RUN --mount=type=cache,target=/root/.cache/go-build \
 
 # dockercli
 FROM base AS dockercli-src
-WORKDIR /tmp/dockercli
-RUN git init . && git remote add origin "https://github.com/docker/cli.git"
-ARG DOCKERCLI_VERSION
-RUN git fetch -q --depth 1 origin "${DOCKERCLI_VERSION}" +refs/tags/*:refs/tags/* && git checkout -q FETCH_HEAD
-RUN [ -d ./components/cli ] && mv ./components/cli /usr/src/dockercli || mv /tmp/dockercli /usr/src/dockercli
 WORKDIR /usr/src/dockercli
+COPY hack/dockerfile/cli.sh ./download-or-build-cli.sh
+ARG DOCKERCLI_REPOSITORY
+RUN git init . && git remote add origin ${DOCKERCLI_REPOSITORY}
 
 FROM base AS dockercli
 WORKDIR /go/src/github.com/docker/cli
 ARG DOCKERCLI_VERSION
-ARG DOCKERCLI_CHANNEL=stable
 ARG TARGETPLATFORM
-RUN xx-apt-get install -y --no-install-recommends gcc libc6-dev
 RUN --mount=from=dockercli-src,src=/usr/src/dockercli,rw \
-    --mount=type=cache,target=/root/.cache/go-build,id=dockercli-build-$TARGETPLATFORM <<EOT
-  set -e
-  DOWNLOAD_URL="https://download.docker.com/linux/static/${DOCKERCLI_CHANNEL}/$(xx-info march)/docker-${DOCKERCLI_VERSION#v}.tgz"
-  if curl --head --silent --fail "${DOWNLOAD_URL}" 1>/dev/null 2>&1; then
-    mkdir /build
-    curl -Ls "${DOWNLOAD_URL}" | tar -xz docker/docker
-    mv docker/docker /build/docker
-  else
-    CGO_ENABLED=0 xx-go build -o /build/docker ./cmd/docker
-  fi
-  xx-verify /build/docker
-EOT
+    --mount=type=cache,target=/root/.cache/go-build,id=dockercli-build-$TARGETPLATFORM \
+    ./download-or-build-cli.sh ${DOCKERCLI_VERSION} /build
+
+FROM base AS dockercli-integration
+WORKDIR /go/src/github.com/docker/cli
+ARG DOCKERCLI_INTEGRATION_VERSION
+ARG TARGETPLATFORM
+RUN --mount=from=dockercli-src,src=/usr/src/dockercli,rw \
+    --mount=type=cache,target=/root/.cache/go-build,id=dockercli-integration-build-$TARGETPLATFORM \
+    ./download-or-build-cli.sh ${DOCKERCLI_INTEGRATION_VERSION} /build
```

Comment:
One potential issue I see here is that this assumes that the "integration cli" and "cli" can both be built from the same repository. But there may be some caveats;

- If I pass a custom `DOCKERCLI_REPOSITORY` (e.g. when I'm working on a CLI feature in my fork), then the `dockercli-src` is switched to that repository
- I _THINK_ in practice that won't make much of a difference (assuming the "integration-cli" is pointing to a tag, and we download a binary)

However, also considering that;

- in the "happy path" we don't use the git source (at all)
- Technically, "integration-cli" and "cli" could be built from separate repositories

.. I'm wondering if we should

- remove the `dockercli-src` stage
- have separate args for the repositories
- make the `git init . && git remote add origin ${DOCKERCLI_REPOSITORY} ... ` step part of `download-or-build-cli.sh`

WDYT?